pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,188 @@
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName]prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+		podLabels,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(labels)
+			} else {
+				logging.FromContext(ctx).Debugf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(labels)
+	}
+	newlabels, err := c.generateLabels(ctx, pod)
+	if err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate new labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge, err := podGaugeVec.GetMetricWith(newlabels)
+	if err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate new gauge: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge.Set(float64(1))
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("podmetrics").
+		For(&v1.Pod{}).
+		Complete(c)
+	return err
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(ctx context.Context, pod *v1.Pod) (prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[podName] = pod.GetName()
+	metricLabels[podNameSpace] = pod.GetNamespace()
+	// Selflink has been deprecated after v.1.20
+	// Manually generate the selflink for the first owner reference
+	// Currently we do not support multiple owner references
+	selflink := ""
+	if len(pod.GetOwnerReferences()) > 0 {
+		ownerreference := pod.GetOwnerReferences()[0]
+		selflink = fmt.Sprintf("/apis/%s/namespaces/%s/%ss/%s", ownerreference.APIVersion, pod.Namespace, strings.ToLower(ownerreference.Kind), ownerreference.Name)
```

Comment:
Was the extra `s` here intentional?
```suggestion
		selflink = fmt.Sprintf("/apis/%s/namespaces/%s/%s/%s", ownerreference.APIVersion, pod.Namespace, strings.ToLower(ownerreference.Kind), ownerreference.Name)
```

Reply:
Additionally, what do you think about creating this as a constant? We may want to use it elsewhere.

```
selfLinkFormatString := "/apis/%s/namespaces/%s/%ss/%s"
```

Reply:
> Was the extra `s` here intentional?

Yes. very hacky way to generate the selflink. In the selflink they want sometime like "replicasets" but the `ownerreference.Kind = "replicaset"`. I am not 100% happy about this. 

> Additionally, what do you think about creating this as a constant? We may want to use it elsewhere.
I think this is really cool idea. And we should really have a helper function to do that.
But this can be potentially controversial since K8S deprecated selflink. 
Can somebody educate me why they decided to deprecate selflink?

Reply:
This pluralization is one of the awkward bits of the kubernetes API. It's an english semantic that only holds true in some cases, e.g. "sheep" -> "sheeps" 😭 . It maybe one of the reasons that Kubernetes deprecated selflink, since it is one of the rare places where this pluralization is used. 

Instead of relying on ownerreferences or jsonified-labels (which both seems like it's a bit awkward here for different reasons), we could instead support https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels like `app.kubernetes.io/name`, `app.kubernetes.io/part-of`. I'm not sure we need to necessarily be in the business of perfectly tying together all possible connections of concepts.